placeholder/src/main/java/com/google/accompanist/placeholder/Placeholder.kt

```diff
@@ -99,85 +119,117 @@ fun Modifier.placeholder(
     val lastSize = remember { Ref<Size>() }
     val lastLayoutDirection = remember { Ref<LayoutDirection>() }
     val lastOutline = remember { Ref<Outline>() }
 
-    // The current animation progress
-    var progress: Float by remember { mutableStateOf(0f) }
+    // The current highlight animation progress
+    var highlightProgress: Float by remember { mutableStateOf(0f) }
+
+    // This is our crossfade transition
+    val transitionState = remember { MutableTransitionState(visible) }.apply {
+        targetState = visible
+    }
+    val transition = updateTransition(transitionState, "placeholder_crossfade")
 
-    // Run the optional animation spec and update the progress
-    progress = highlight?.animationSpec?.let { spec ->
+    val placeholderAlpha by transition.animateFloat(
+        transitionSpec = placeholderFadeTransitionSpec,
+        label = "placeholder_fade",
+        targetValueByState = { placeholderVisible -> if (placeholderVisible) 1f else 0f }
+    )
+    val contentAlpha by transition.animateFloat(
+        transitionSpec = contentFadeTransitionSpec,
+        label = "content_fade",
+        targetValueByState = { placeholderVisible -> if (placeholderVisible) 0f else 1f }
+    )
+
+    // Run the optional animation spec and update the progress if the placeholder is visible
+    val animationSpec = highlight?.animationSpec
+    if (animationSpec != null && (visible || placeholderAlpha >= 0.01f)) {
         val infiniteTransition = rememberInfiniteTransition()
-        infiniteTransition.animateFloat(
+        highlightProgress = infiniteTransition.animateFloat(
             initialValue = 0f,
             targetValue = 1f,
-            animationSpec = spec,
+            animationSpec = animationSpec,
         ).value
-    } ?: 0f
+    }
 
+    val paint = remember { Paint() }
     remember(color, shape, highlight) {
         drawWithContent {
-            // Draw normal composable content first
-            drawContent()
+            // Draw the composable content first
+            if (contentAlpha in 0.01f..0.99f) {
+                // If the content alpha is between 1% and 99%, draw it in a layer with
+                // the alpha applied
+                paint.alpha = contentAlpha
+                withLayer(paint) {
+                    with(this@drawWithContent) {
+                        drawContent()
+                    }
+                }
+            } else if (contentAlpha >= 0.99f) {
+                // If the content alpha is > 99%, draw it with no alpha
+                drawContent()
+            }
 
-            if (shape === RectangleShape) {
-                // shortcut to avoid Outline calculation and allocation
-                // Draw the initial background color
-                drawRectPlaceholder(
-                    color = color,
-                    highlight = highlight,
-                    progress = progress
-                )
-            } else {
+            fun drawPlaceholder() {
```

Comment:
I would consider defining this function outside of the draw block as this will allocate on each frame.

Reply:
Thanks, fixed in 350f4b99fe9c04a074c9e39e25cf0eb7786ef2d7